pyro/distributions/improper_uniform.py

```diff
@@ -0,0 +1,67 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+from torch.distributions import constraints
+
+from .torch_distribution import TorchDistribution
+from .util import broadcast_shape
+
+
+class ImproperUniform(TorchDistribution):
+    """
+    Improper distribution with zero :meth:`log_prob` and undefined
+    :meth:`sample`.
+
+    This is useful for transforming a model from generative dag form to factor
+    graph form for use in HMC. For example the following are equal in
+    distribution::
+
+        # Version 1. a generative dag
+        x = pyro.sample("x", Normal(0, 1))
+        y = pyro.sample("y", Normal(x, 1))
+        z = pyro.sample("z", Normal(y, 1))
+
+        # Version 2. a factor graph
+        xyz = pyro.sample("xyz", ImproperUniform(constraints.real, (), (3,)))
+        x, y, z = xyz.unbind(-1)
+        pyro.sample("x", Normal(0, 1), obs=x)
+        pyro.sample("y", Normal(x, 1), obs=y)
+        pyro.sample("z", Normal(y, 1), obs=z)
+
+    Note this distribution errors when :meth:`sample` is called. To create a
+    similar distribution that instead samples from a specified distribution
+    consider using ``.mask(False)`` as in::
+
+        xyz = dist.Normal(0, 1).expand([3]).to_event(1).mask(False)
+
+    :param support: The support of the distribution.
+    :type support: ~torch.distributions.constraints.Constraint
+    :param torch.Size batch_shape: Defaults to empty.
+    :param torch.Size event_shape: Defaults to empty.
+    """
+    arg_constraints = {}
+
+    def __init__(self, support, batch_shape=torch.Size(), event_shape=torch.Size()):
```

Comment:
I think `event_shape` is a requirement. Otherwise, we can set it to `None`, so that the default event_shape can be `(0,) * biject_to(support).event_dim`. This way, log_prob will have a more appropriate shape.

Reply:
I think i'd rather make them both required than perform a computation on the constraint. Currently our constraints are pretty sloppy- we should be using `IndependentConstraint` more thoroughly and many of the computations would be wrong. This isn't a big issue yet because we don't use `biject_to(support).event_dim` for much.

Reply:
Yeah, I agree that we should require event_shape. But for batch_shape, I guess it is better to let it have default shape as before (i.e. `__init__(self, support, event_shape, batch_shape=torch.Size())`) to makes the code less verbose.

Reply:
I agree a default of `batch_shape=()` makes sense, but I find it disorienting for batch_shape to be specified on the right of event_shape: I always see things as `batch_shape, event_shape`. 🤷

Reply:
Haha, sounds reasonable to me.